src/serial_io.rs

```diff
@@ -0,0 +1,45 @@
+use heapless::spsc::Queue;
+use esp32c3_hal::{ 
+    Serial,
+    pac::UART0,
+};
+use esp_hal_common::{ 
+    serial::Instance, 
+    interrupt,
+    interrupt::CpuInterrupt::*,
+    Cpu::*,
+};
+use crate::protocol::InputIO;
+use crate::targets::esp32c3 as target;
+
+const RX_QUEUE_SIZE: usize = target::MAX_WRITE_BLOCK + 0x400;
+static mut RX_QUEUE: Queue<u8, RX_QUEUE_SIZE> = Queue::new();
+
+impl<'a, T: Instance> InputIO for Serial<T> {
+    fn recv(&mut self) -> u8 {
+        let mut consumer = unsafe { RX_QUEUE.split().1 };
+
+        unsafe{
+            while consumer.ready() == false { }
+            consumer.dequeue_unchecked()
+        }
+    }
+
+    fn send(&mut self, bytes: &[u8]) {
+        self.write_bytes(bytes).unwrap()
+    }
+}
+    
+#[no_mangle]
+pub fn interrupt3() {
+    let uart = unsafe{ &*UART0::ptr() };
+    let mut producer = unsafe { RX_QUEUE.split().0 };
+    
+    while uart.status.read().rxfifo_cnt().bits() > 0 {
+        let data = uart.fifo.read().rxfifo_rd_byte().bits();
+        unsafe{ producer.enqueue_unchecked(data) };
```

Comment:
Why is this unchecked? We may miss data if we don't check it. Even if we can't recover from having a full queue, we should at least throw an error else we may run into some tricky to debug situations in the future.

Reply:
I've changed this to call the `enqueue()` function instead, but wasn't entirely sure what to do with the `Result`. I'm just unwrapping it, but if there's something else we should do instead let me know and I'm happy to change it.

Reply:
`unwrap` is fine I think, better to panic than silently miss data imo